### ISSUE TITLE
Source to Image Compatibility

### ISSUE THREAD

**Issue by sabre1041** [sclorg/httpd-container#7]
Since the `USER` instruction is currently commented out, the image cannot be used with the Source to Image mechanism

**Comment by bparees** [comment]
@hhorak I assume there's a plan to fix this as part of delivering the s2i enabled SCL httpd image?

openshift does not allow s2i builder images to run as root, so it's necessary they default to a numeric non-zero uid, eg: https://github.com/sclorg/s2i-ruby-container/blob/master/2.3/Dockerfile#L33

**Comment by hhorak** [comment]
httpd has been running as root by default for some time already, so the idea behind s2i support was that once a new image is produced by `s2i build ...`, this new image will be usable as non-root. But this makes it still unusable in OpenShift, since it is not allowed to run even s2i tool as root, right?
Btw. running s2i as root would allow us to support much more use cases, like installing some database plugin from RPM, some more packages available as RPMs for python/ruby... Even if it wouldn't be allowed in OpenShift, I think having "run as root" option in s2i would be still usable outside of openshift..

**Comment by bparees** [comment]
Correct, the "assemble" script cannot be run as root, because we do not allot users to run anything as root inside the openshift cluster, and running the assemble script as root would allow them to provide arbitrary user commands in the form of a custom assemble script, and run them as root inside a container running in openshift.

as for s2i as a general tool allowing this, well, it does.  if you do an s2i build operations on the CLI, it does not care if the assemble script runs as root.  But of course you can't use that same image in openshift, then.

**Comment by jwforres** [comment]
Is it possible to get this issue fixed? I am having to create a new builder image based on this one that sets `USER 1001` so that I can do S2I builds on openshift.  We are trying to base the openshift console's new runtime image on this builder image and I would prefer to depend on it directly instead of having the custom image to change the user.

**Comment by hhorak** [comment]
I totally understand that having and building another Dockerfile just in order to run the httpd container is not good, on the other hand I still see reasons why some users might want to run httpd as root user (separating main process from child processes for security reasons; both running as different users).

However, since OpenShift use case is important here, I'd say we can change default user and properly document how to run the container as root. With changing the root we also loose option to run apache on 80/443 ports by default inside the containers, which in my opinion shouldn't cause too much troubles, users would just need to change the port mapping in this case. Yes, these are incompatible changes, but IMO still worth it. So, what we definitely need to do is proper documentation.

Since this is change in default behavior, let me check first, what all we need to take care of..

**Comment by dlbewley** [comment]
I could be missing something, but it seems like `USER 1001` is the expected default. 

I'm able to make the nginx s2i available to my users by just creating an imagestream. I'd love to be able to do the same with httpd and not have to maintain another builder image.

Any hope this will get _fixed_ soon?

**Comment by hhorak** [comment]
I've finally prepared the fix for this issue, but since it is incompatible, we'll consider it for RHSCL 2.4 update, so we'll be able to document it properly. Since there were more issues with how the code worked, in the end I decided to re-factor the whole sources to make them more maintainable and to offer more features. All the work is now available in #13.

**Comment by dlbewley** [comment]
Should we expect to see this upstream now? I don't seem to.

```bash
$ docker run -ti --rm --entrypoint=grep registry.access.redhat.com/rhscl/httpd-24-rhel7:latest USER /root/buildinfo/Dockerfile-rhscl-httpd-24-rhel7-2.4-32.11
# USER 1001
```

**Comment by hhorak** [comment]
I think we can close this issue now, the downstream delivery is tracked in https://bugzilla.redhat.com/show_bug.cgi?id=1362197
Closed by #13.
